src/wp-includes/fonts/class-wp-font-family.php

```diff
@@ -0,0 +1,615 @@
+	 */
+	private function get_intersecting_font_faces( $existing, $incoming ) {
+		$intersecting = array();
+		foreach ( $existing as $existing_face ) {
+			foreach ( $incoming as $incoming_face ) {
+				if ( $incoming_face['fontStyle'] === $existing_face['fontStyle'] &&
+					$incoming_face['fontWeight'] === $existing_face['fontWeight'] &&
+					$incoming_face['src'] !== $existing_face['src'] ) {
+					$intersecting[] = $existing_face;
+				}
+			}
+		}
+		return $intersecting;
+	}
+
+	/**
+	 * Updates a post for a font family.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param WP_Post $post The post to update.
+	 * @return int|WP_Error Post ID if the update was successful, WP_Error otherwise.
+	 */
+	private function update_font_post( $post ) {
+		$post_font_data = json_decode( $post->post_content, true );
+		$new_data       = WP_Font_Family_Utils::merge_fonts_data( $post_font_data, $this->data );
+		if ( isset( $post_font_data['fontFace'] ) && ! empty( $post_font_data['fontFace'] ) ) {
+			$intersecting = $this->get_intersecting_font_faces( $post_font_data['fontFace'], $new_data['fontFace'] );
+		}
+
+		if ( isset( $intersecting ) && ! empty( $intersecting ) ) {
+			$serialized_font_faces   = array_map( 'serialize', $new_data['fontFace'] );
+			$serialized_intersecting = array_map( 'serialize', $intersecting );
+
+			$diff = array_diff( $serialized_font_faces, $serialized_intersecting );
+
+			$new_data['fontFace'] = array_values( array_map( 'unserialize', $diff ) );
+
+			foreach ( $intersecting as $intersect ) {
+				$this->delete_font_face_assets( $intersect );
+			}
+		}
+		$this->data = $new_data;
+
+		$post = array(
+			'ID'           => $post->ID,
+			'post_content' => $this->get_data_as_json(),
+		);
+
+		$post_id = wp_update_post( $post );
+
+		if ( 0 === $post_id || is_wp_error( $post_id ) ) {
+			return new WP_Error(
+				'font_post_update_failed',
+				__( 'Font post update failed' )
+			);
+		}
+
+		return $post_id;
+	}
+
+	/**
+	 * Creates a post for a font in the Font Library if it doesn't exist,
+	 * or updates it if it does.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return int|WP_Error Post id if the post was created or updated successfully,
+	 *                      WP_Error otherwise.
+	 */
+	private function create_or_update_font_post() {
+		$this->sanitize();
+
+		$post = $this->get_font_post();
+		if ( $post ) {
+			return $this->update_font_post( $post );
+		}
+
+		return $this->create_font_post();
+	}
+
+	/**
+	 * Installs the font family into the library.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $files Optional. An array of files to be installed. Default null.
+	 * @return array|WP_Error An array of font family data on success, WP_Error otherwise.
+	 */
+	public function install( $files = null ) {
+		add_filter( 'upload_mimes', array( 'WP_Font_Library', 'set_allowed_mime_types' ) );
+		add_filter( 'upload_dir', array( 'WP_Font_Library', 'set_upload_dir' ) );
```

Comment:
How about sites running s3 upload?

Reply:
I didn't try with S3 plugin yet, but I think it's out of the scope of this PR, right?

Reply:
cc @azaozz